### ISSUE TITLE
rewrite recursive conditional types to be tail-recursive

### ISSUE THREAD

**Issue by DetachHead** [millsp/ts-toolbelt#255]
#### Is your feature request related to a problem?

many types in ts toolbelt use recursive conditional types, meaning that using them often results in stack depth errors.
#### Describe the solution you'd like

in typescript 4.5, tail-recursive conditional types are now optimized such that the recursion limit can be much higher (1000) instead of 100. see the [blog post](https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-beta/#tailrec-conditional) and https://github.com/microsoft/TypeScript/pull/45711

**Comment by millsp** [comment]
That would be an awesome addition :) Have you evaluated which types aren't working out of the box?

**Comment by DetachHead** [comment]
ones i've noticed so far (will update this list as i come across more):
- `String/Join`
- `List/Flatten`
- `IterationMap`

**Comment by DetachHead** [comment]
i'm not sure how to go about updating `IterationMap`, i don't want to just add 1000 more entries to it because that seems gross. perhaps there's a better solution?

**Comment by millsp** [comment]
> i'm not sure how to go about updating it, i don't want to just add 1000 more entries to it because that seems gross. perhaps there's a better solution?

That is totally fine. The map itself is not dynamic on purpose and is there to optimize TypeScript to make iteration convenient and efficient. In essence, it's there to teach TypeScript how to count. The `Next` and `Prev` operators operate through it in the most efficient manner.

**Comment by millsp** [comment]
Yes, it's because you can index with `0` and `1` but not `true` and `false`. The reason for this is that `extends` clauses are way more expensive than a deterministic index. That means that the compiler is able to flatten such types faster that with an `extends` clause.

**Comment by millsp** [comment]
> however from what i understand the optimization results in it evaluating the type more efficiently anyway

I also think so. That means that this will be quite an amount of work. In the entry point of the lib, I provided a regex for finding all the recursive types - that will give us an idea of how much there is to do. I'll be happy to contribute on this PR with you :)
